src/Transport/Receiving/MessagePump.cs

```diff
@@ -322,33 +325,56 @@ await processor.CloseAsync(cancellationToken)
                                 transportSettings.TransportTransactionMode,
-                                cancellationToken: messageProcessingCancellationToken)
+                                cancellationToken: processingTokenSource.Token)
                             .ConfigureAwait(false);
                     }
                 }
-                catch (ServiceBusException onErrorEx) when (onErrorEx.IsTransient || onErrorEx.Reason is ServiceBusFailureReason.MessageLockLost)
+                catch (ServiceBusException onErrorEx) when (onErrorEx.IsTransient ||
+                                                            onErrorEx.Reason is ServiceBusFailureReason.MessageLockLost)
                 {
                     Logger.Debug("Failed to execute recoverability.", onErrorEx);
 
                     await processMessageEventArgs.SafeAbandonMessageAsync(message,
                             transportSettings.TransportTransactionMode,
-                            cancellationToken: messageProcessingCancellationToken)
+                            cancellationToken: processingTokenSource.Token)
                         .ConfigureAwait(false);
                 }
-                catch (Exception onErrorEx) when (onErrorEx.IsCausedBy(messageProcessingCancellationToken))
+                catch (Exception onErrorEx) when (onErrorEx.IsCausedBy(processingTokenSource.Token))
                 {
                     throw;
                 }
                 catch (Exception onErrorEx)
                 {
-                    criticalErrorAction($"Failed to execute recoverability policy for message with native ID: `{message.MessageId}`", onErrorEx, messageProcessingCancellationToken);
+                    // Using messageProcessingCancellationToken to make sure the critical error action can continue
+                    // to execute until the message processing is stopped.
+                    criticalErrorAction(
+                        $"Failed to execute recoverability policy for message with native ID: `{message.MessageId}`",
+                        onErrorEx, messageProcessingCancellationToken);
 
                     await processMessageEventArgs.SafeAbandonMessageAsync(message,
                             transportSettings.TransportTransactionMode,
-                            cancellationToken: messageProcessingCancellationToken)
+                            cancellationToken: processingTokenSource.Token)
                         .ConfigureAwait(false);
                 }
             }
+            finally
+            {
+                processMessageEventArgs.MessageLockLostAsync -= MessageLockLostHandler;
+            }
+
+            async Task MessageLockLostHandler(MessageLockLostEventArgs lockLostArgs)
+            {
+                Logger.Info("Lost the lock while processing message. Cancelling the pipeline execution.", lockLostArgs.Exception);
```

Comment:
I am unsure what is the benefit of adding the exception to the log in this log entry?

Reply:
If exception contains the message ID it helps. But, this message is not NSB message ID. And because physical message ID changes on delayed retries, it's less helpful during troubleshooting.

So I'd say maybe log NSB message ID.

Reply:
The pump does enforce the message id to be set currently and throws otherwise. The message ID is already logged in a few places so I extended and aligned it here too.